include/zephyr/usb/class/usb_midi.h

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2024 Titouan Christophe
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_USB_CLASS_USB_MIDI_H_
+#define ZEPHYR_INCLUDE_USB_CLASS_USB_MIDI_H_
+
+/**
+ * @brief USB-MIDI 2.0 class device API
+ * @defgroup usb_midi USB MIDI 2.0 Class device API
+ * @ingroup usb
+ * @since 4.1
+ * @version 0.1.0
+ * @see ump112: "Universal MIDI Packet (UMP) Format and MIDI 2.0 Protocol"
+ *              Document version 1.1.2
+ * @{
+ */
+
+#include <zephyr/device.h>
+
+/**
+ * @defgroup usb_midi_mt Universal MIDI Packet message types
+ * @ingroup usb_midi
+ * @see ump112: 2.1.4 Message Type (MT) Allocation
+ * @{
+ */
+
+#define MT_UTILITY             0x00
+#define MT_SYS_RT_COMMON       0x01
+#define MT_MIDI1_CHANNEL_VOICE 0x02
+#define MT_DATA                0x03
+#define MT_MIDI2_CHANNEL_VOICE 0x04
+#define MT_FLEX_DATA           0x0d
+#define MT_UMP_STREAM          0x0f
+
+/**
+ * @}
+ *
+ * @brief      Size of a Universal MIDI Packet, in 32bit words
+ * @param[in]  mt    The packet Message Type
+ * @see ump112: 2.1.4 Message Type (MT) Allocation
+ */
+static inline size_t ump_words(uint8_t mt)
+{
+	return ((uint8_t[16]) {1, 1, 1, 2, 2, 4, 1, 1, 2, 2, 2, 3, 3, 4, 4, 4})[mt & 0x0f];
+}
+
+
+/**
+ * @brief      MIDI1 channel voice message in Universal MIDI Packet format
+ * @see ump112: 7.3 MIDI 1.0 Channel Voice Messages
+ */
+struct ump_midi1 {
+#ifdef CONFIG_LITTLE_ENDIAN
+	/** @brief The 2nd MIDI1 parameter */
+	uint8_t p2;
+	/** @brief The 1st MIDI1 parameter */
+	uint8_t p1;
+	/** @brief The MIDI1 channel number */
+	unsigned channel: 4;
+	/** @brief The MIDI1 status nibble (command) */
+	unsigned status: 4;
+	/** @brief The UMP group to which this message belongs */
+	unsigned group: 4;
+	/** @brief The UMP message type (MT_MIDI1_CHANNEL_VOICE) */
+	unsigned mt: 4;
+#else
+	/** @brief The UMP message type (MT_MIDI1_CHANNEL_VOICE) */
+	unsigned mt: 4;
+	/** @brief The UMP group to which this message belongs */
+	unsigned group: 4;
+	/** @brief The MIDI1 status nibble (command) */
+	unsigned status: 4;
+	/** @brief The MIDI1 channel number */
+	unsigned channel: 4;
+	/** @brief The 1st MIDI1 parameter */
+	uint8_t p1;
+	/** @brief The 2nd MIDI1 parameter */
+	uint8_t p2;
+#endif
```

Comment:
I leave it up to you to decide whether to use a macro to set the bitfields on `uint32_t` here or rely on compiler putting things in correct order. The bit order is in no way guaranteed by the standard. I do know that parts of USB stack use this approach, but I don't quite like it.

Reply:
Yes indeed, I don't know what's the best way of defining midi packet fields here. I got some inspiration from other USB structs (for example https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/usb/usb_ch9.h#L27-L49).

While I like being able to access individual fields from the midi packet as a struct, this is tedious and error prone, because all fields have to be repeated in the correct order twice.

Not sure if there is some kind of convention to follow here ? I'd happily change that part

Reply:
I don't know if there is any specific convention for this type of things. The portable code to handle such things is to have macros to put fields on correct bit places. I am not really familiar with actual MIDI usage so cannot really come up with what sort of macros would work best.

Reply:
I just pushed a new version (rebased on the latest main) that replace these enums with accessors/constructor macros on `uint32_t []`